### PR TITLE
Fix memory leak in torch._dirichlet_grad()

### LINKED ISSUE #1853: Memory leak using Dirichlet distribution in guide
### Issue Description
Memory leak encountered on CPU when sampling from a Dirichlet distribution in a guide over many training iterations. The leak only occurs when the parameters of the Dirichlet distribution are defined using `param`.

### Environment
OS: Ubuntu 18.04.2 
Python: 3.6.7 
PyTorch: 1.1.0 
Pyro: 0.3.3

### Code Snippet
```py
import torch
from torch.distributions import constraints

import pyro
from pyro.infer import SVI, Trace_ELBO
from pyro.optim import Adam
import pyro.distributions as dist


a = 20
b = 100000

def model(x):
    pass

def guide(x):
    alpha = pyro.param('alpha', torch.ones(a, b), constraint=constraints.positive)                  
    with pyro.plate('sample_x', a):
        pyro.sample('x', dist.Dirichlet(alpha))

elbo = Trace_ELBO(max_plate_nesting=1, vectorize_particles=True)
opt = Adam({'lr': 0.0001})
svi = SVI(model, guide, opt, loss=elbo)

for epoch in range(500):
    svi.step(1)
```
Changing `alpha` in the guide to 
```
alpha = torch.ones(a, b) / b 
```
stops the leak from occurring (ignore errors).

### PR DESCRIPTION
**Pull request by fritzo**
Fixes https://github.com/pyro-ppl/pyro/issues/1853

This fixes a memory leak in `torch._dirichlet_grad()`. This function is used for reparametrized gradients for the `Dirichlet` and `Beta` distributions.

## Questions for reviewiers

- [x] Could a reviewer please confirm that `freeCopyTo()` is being used correctly and doesn't need an additional `decref()`? The author is unfamiliar with PyTorch C++ memory utilities. Help appreciated.

## Tested
- ran locally and confirmed leak is fixed

### CODE REVIEW

**Comment by fritzo** [comment]
Travis failure is unrelated, due to python lint errors (this PR only touches C++ code).

**Review by vishwakftw — APPROVED** [review]
I think the fix is right. `freeCopyTo` copies the value in `grad` to `self` and then deallocates `grad`.
